hypothesis-python/tests/cover/test_database_backend.py

```diff
@@ -500,3 +509,245 @@ def test_uleb_128_roundtrips(n1):
+
+        def _expect_delete(self, k, v):
+            if not supports_value_delete:
+                v = None
+            self._expect_event("delete", (k, v))
+
+        def _expect_save(self, k, v):
+            self._expect_event("save", (k, v))
+
+        def _expect_move(self, k1, k2, v):
+            self._expect_event("move", (k1, k2, v))
+
+        @rule(target=keys, k=st.binary())
+        def k(self, k):
+            return k
+
+        @rule(target=values, v=st.binary())
+        def v(self, v):
+            return v
+
+        @precondition(lambda self: not self.active_listeners)
+        @rule()
+        def add_listener(self):
+            self.db.add_listener(self.listener)
+            self.active_listeners.append(self.listener)
+
+        @precondition(lambda self: self.listener in self.active_listeners)
+        @rule()
+        def remove_listener(self):
+            self.db.remove_listener(self.listener)
+            self.active_listeners.remove(self.listener)
+
+        @rule()
+        def clear_listeners(self):
+            self.db.clear_listeners()
+            self.active_listeners.clear()
+
+        @rule(k=keys)
+        def fetch(self, k):
+            # we don't expect this to do anything, but that's the point. if this
+            # fires a listener call then that's bad and will fail.
+            self.db.fetch(k)
+
+        @rule(k=keys, v=values)
+        def save(self, k, v):
+            changed = v not in set(self.db.fetch(k))
+            self.db.save(k, v)
+
+            if changed:
+                self._expect_save(k, v)
+
+        @rule(k=keys, v=values)
+        def delete(self, k, v):
+            changed = v in set(self.db.fetch(k))
+            self.db.delete(k, v)
+
+            if changed:
+                self._expect_delete(k, v)
+
+        @rule(k1=keys, k2=keys, v=values)
+        def move(self, k1, k2, v):
+            in_k1 = v in set(self.db.fetch(k1))
+            save_changed = v not in set(self.db.fetch(k2))
+            delete_changed = k1 != k2 and in_k1
+            self.db.move(k1, k2, v)
+
+            # in database backends that support it, this gets emitted as a move
+            # (*unless* the keys are equal, in which case it is a save).
+            #
+            # otherwise it gets emitted as a delete followed by a save, unless
+            # those did not change the db.
+            if supports_move:
+                if k1 == k2 and save_changed:
+                    self._expect_save(k2, v)
+                elif save_on_no_k1_in_move and not in_k1 and save_changed:
+                    self._expect_save(k2, v)
+                elif delete_changed or save_changed:
+                    self._expect_move(k1, k2, v)
+            else:
+                if delete_changed:
+                    self._expect_delete(k1, v)
+                if save_changed:
+                    self._expect_save(k2, v)
+
+        @invariant()
```

Comment:
I'd be tempted to make this a `@rule()`, so that we can test multiple operations without flushing in between.

Reply:
I actually had this but ran into race conditions where a listener was removed immediately after a `save/delete/move` operation, before the listener could fire - then when it did fire, there was no listener to send it to. (only relevant for the directory database, I guess)